core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -133,6 +134,14 @@ class LogCleaner(initialConfig: CleanerConfig,
              def value: Int = cleaners.map(_.lastStats).map(_.elapsedSecs).max.toInt
            })
 
+  newGauge("log-cleaner-dead-thread-count",
+    new Gauge[Int] {
+      def value: Int = deadThreadCount
+    })
+
+  // visible for testing
+  private[log] def deadThreadCount = cleaners.count(_.isShutdownComplete)
```

Comment:
I think we should distinguish between threads that have shutdown intentionally and ones that have errored out. I'm afraid we may have race conditions where this gauge != 0 yet logically it is.
See the `LogCleaner#reconfigure()` method. It calls `LogCleaner#shutdown()` which does 
```
cleaners.foreach(_.shutdown())
cleaners.clear()
```
Meaning we could catch some of these cleaners with `isShutdownComplete` before we've cleared the `cleaners` array.

In such cases we wouldn't want to report any dead threads, I don't think. Maybe we can also check for the threads' `shutdownInitiated` countdown latch to differentiate between intentional shutdowns and not

Reply:
Yea, that's a good idea, we could use the `shutdownComplete.getCount == 0 && shutdownInitiated.getCount != 0` construct or something like this. I have doubt with this though:
```
} catch {
      case e: FatalExitError =>
        shutdownInitiated.countDown()
        shutdownComplete.countDown()
        info("Stopped")
        Exit.exit(e.statusCode())
      case e: Throwable =>
        if (isRunning)
          error("Error due to", e)
    }
```
It just might be slightly confusing at first because we're counting down on both but this is an unintentional shutdown. But in this case the whole broker will terminate so it wouldn't even matter.

Reply:
I think so too